BaseStation/VideoStreamer/MainWindow.xaml.cs

```diff
@@ -1,29 +1,231 @@
+        }
+
+        // plays the video streams in their own windows
+        private void ButtonLaunch(object sender, RoutedEventArgs e)
+        {
+            if (playing)
+            {
+                playing = false;
+                btnLaunch.Content = "Launch";
+                closeAll(players);
+            }
+            else
+            {
+                playing = true;
+                btnLaunch.Content = "Stahp";
+                initProcesses(ref players);
+                for (int i = 0; i < ports.Count; i++)
+                {
+                    startStream(ports.ElementAt(i), players[i]);
+                }
+            }
+        }
+
+        // closes all windows in the array of processes given
+        private void closeAll(Process[] cmd)
+        {
+            for (int i = 0; i < cmd.Length; i++)
+            {
+                cmd[i].CloseMainWindow();
+            }
+        }
+
+        // initiates a stream with a given port
+        private void startStream(int port, Process cmd)
+        {
+            cmd.StartInfo.FileName = "cmd.exe";
+            cmd.StartInfo.RedirectStandardInput = true;
+            cmd.StartInfo.UseShellExecute = false;
+            cmd.Start();
+            cmd.StandardInput.WriteLine(@"set PATH=%PATH%;C:\gstreamer\1.0\x86_64\bin");
+            cmd.StandardInput.WriteLine("gst-launch-1.0 -vvv udpsrc caps=\"application/x-rtp," +
+                " media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264\" port=" +
+                port + " ! rtph264depay ! decodebin ! videoconvert ! autovideosink");
+        }
+
+        // saves the stream to a file from the given port, destination is set from the control panel
+        // and a pseudo-random name is generated for the file using the current date and time
+        private void recordStream(int port, Process cmd, int cam)
+        {
+            cmd.StartInfo.WorkingDirectory = Destination;
+            cmd.StartInfo.FileName = @"C:\gstreamer\1.0\x86_64\bin\gst-launch-1.0.exe";
+            cmd.StartInfo.Arguments = "-m -vvv -e udpsrc caps=\"application/x-rtp," +
+                " media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264\" port=" +
+                port + " ! rtph264depay ! h264parse ! mp4mux ! filesink location=" +
+                DateTime.Now.ToString("MM-dd-yyyy--HH;mm;ss") + "(" + cam + ")" + ".mp4";
+            cmd.Start();
+        }
+
+        // adds ports to the port list
+        private void Add(object sender, RoutedEventArgs e)
+        {
+            int x;
+            if (Int32.TryParse(portBox.Text, out x))
+            {
+                Ports.Add(x);
+            }
+            portBox.Text = "";
+        }
+
+        // removes selected items from the port list
+        private void Remove(object sender, RoutedEventArgs e)
+        {
+            List<int> remove = new List<int>();
+            foreach (int thing in portList.SelectedItems)
```

Comment:
Will List.RemoveAll work here? It's more concise and removes the need for this intermediate list.

https://stackoverflow.com/questions/3279145/remove-item-from-list-based-on-condition

Reply:
I've tried that and running it creates an error saying list may have changed and refuses to continue

Reply:
That's strange. It's probably fine then.